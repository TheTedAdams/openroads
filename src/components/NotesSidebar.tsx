import * as React from 'react';
import { INote, INotesSideProps } from "../types/types"

import classNames from 'classnames';
import checkMark from '../images/check-mark.svg';

class NotesSidebar extends React.Component<INotesSideProps> {
  getNotesRows() {
    // fixed: selected row highlight
    const { currentIndex, notes, onSelectNote } = this.props 
    return notes.map((note: INote) => (
      <div
        key={note.subject}
        className={classNames('NotesSidebarItem', {
          selected: notes.indexOf(note).toString() === currentIndex.toString() 
        })}
        onClick={onSelectNote}
        id={notes.indexOf(note).toString()}
      >
        <h4 className='NotesSidebarItem-title'>{note.subject}</h4>
        {note.read && <img alt='Check Mark' src={checkMark} />}
      </div>
    ));
  }
  render() {
    return(
      <section className='NotesSidebar'>
        {/* fixed: layout of notesSidebar, left of the NoteDetails, taking up
        the full height of the space beneath the header. */}
        <h2 className='NotesSidebar-title'>Available Notes:</h2>
        <div>{this.getNotesRows()}</div>
      </section>
    );
  }
}

export default NotesSidebar;