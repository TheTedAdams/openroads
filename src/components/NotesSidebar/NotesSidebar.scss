@import '../../styles/colors';

.NotesSidebar {
  grid-column: 1;
  grid-row: 2;
  background: $brown;
  border-right: solid 1px $gray;
  color: white;

  p {
    margin: 0.3rem;
  }

  &-title {
    margin: 0;
    font-size: 1.25rem;
    padding: 0.2rem;
  }

  &Item {
    padding: 0 0.3rem;
    display: flex;
    flex-direction: row;
    border-style: solid;
    border-color: $gray;
    border-width: 1px 0 0;
    align-items: center;

    // Button reset stuff
    background: none;
    width: 100%;
    text-align: left;
    font-size: inherit;
    color: inherit;

    &.selected {
      background: $yellow;
    }

    &:not(.selected):hover {
      background: rgba($yellow, 0.4);
    }

    &-title {
      margin: 0.3rem 0;
      display: inline-block;
      flex: 1;
    }
  }
}
