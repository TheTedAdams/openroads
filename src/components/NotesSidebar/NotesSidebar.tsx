import './NotesSidebar.scss';

import React, { FC, useCallback } from 'react';
import { Note } from '../../models';
import { NotesSidebarItem } from './parts/NotesSidebarItem';

interface NotesSidebarProps {
  isLoading: boolean;
  notes: Note[];
  onSelectNote: (noteIndex: number) => void;
  currentNoteIndex: number;
}

export const NotesSidebar: FC<NotesSidebarProps> = ({
  isLoading,
  notes,
  onSelectNote,
  currentNoteIndex,
}) => {
  const handleSelect = useCallback((index: number) => onSelectNote(index), [
    onSelectNote,
  ]);

  return (
    <section className="NotesSidebar">
      <h2 className="NotesSidebar-title">Available Notes:</h2>
      <div className="NotesSidebar-list">
        {isLoading ? (
          <p>Loading...</p>
        ) : (
          notes.map((note, index) => (
            <NotesSidebarItem
              note={note}
              isSelected={index === currentNoteIndex}
              index={index}
              onSelect={handleSelect}
            />
          ))
        )}
      </div>
    </section>
  );
};
