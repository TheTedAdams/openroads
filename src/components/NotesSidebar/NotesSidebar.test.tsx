import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import { NotesSidebar } from './NotesSidebar';

const TEST_NOTES = [
  { subject: 'SUBJECT 1', body: 'BODY 1', read: false },
  { subject: 'SUBJECT 2', body: 'BODY 2', read: true },
];

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(
    <NotesSidebar
      isLoading={false}
      onSelectNote={() => null}
      notes={TEST_NOTES}
      currentNoteIndex={0}
    />,
    div
  );
  ReactDOM.unmountComponentAtNode(div);
});

it('Renders like the snapshot', () => {
  const tree = renderer
    .create(
      <NotesSidebar
        isLoading={false}
        onSelectNote={() => null}
        notes={TEST_NOTES}
        currentNoteIndex={0}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
