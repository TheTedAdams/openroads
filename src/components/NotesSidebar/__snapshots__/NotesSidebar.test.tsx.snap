// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders like the snapshot 1`] = `
<section
  className="NotesSidebar"
>
  <h2
    className="NotesSidebar-title"
  >
    Available Notes:
  </h2>
  <div
    className="NotesSidebar-list"
  >
    <button
      className="NotesSidebarItem selected"
      onClick={[Function]}
    >
      <h4
        className="NotesSidebarItem-title"
      >
        SUBJECT 1
      </h4>
    </button>
    <button
      className="NotesSidebarItem"
      onClick={[Function]}
    >
      <h4
        className="NotesSidebarItem-title"
      >
        SUBJECT 2
      </h4>
      <img
        alt="Check Mark"
        src="check-mark.svg"
      />
    </button>
  </div>
</section>
`;
